### ISSUE TITLE
isUnderAttack() behavior 

### ISSUE THREAD

**Issue by albertouri** [bwapi/bwapi#549]
isUnderAttack() is more like itHasBeenAttacked()
For instance, it cannot handle instakills.

It would be better if isUnderAttack() returns true when a unit with a AttackUnit state is targeting you.
Or maybe a new methos isExposed() that returns true if the unit is in the weapon range of an enemy unit.

**Comment by GeEom** [comment]
At some point which frame Starcraft internally specifies as the start of an attack becomes arbitrary.

Consider: a siege tank can see you, and you cannot see it. It may well begin the animation prerequisite for attacking. If you leave range in this animation then specifying 'under attack' would be incorrect. Further, if 'under attack' is set as the tank animates, your client has a premonition on information you don't otherwise observe.
If we presume flags are set on this basis, then polling weapon range, attack cooldown, animation frames, etc manually (for which functionality is provided) does give more information than isUnderAttack.
Determining ASAP what is going on, irrespective of specific flags, has possible overlap with https://github.com/bwapi/bwapi/issues/525.

In this instance however, I don't see how having BWAPI handle it gives advantage over DIY. As the relevant data is so flexible, I'd want an exact understanding of what beingAttacked?()==true actually implies.

**Comment by krasi0** [comment]
Your analysis is correct. I have implemented a DIY solution which iterates
over all enemy units that have the specific unit in weapon range and are
aiming at the unit. Slower but more accurate.

**Comment by heinermann** [comment]
Right now isUnderAttack just returns true if some internal UI timer was set indicating that the unit was recently attacked. There is no logic behind it in BWAPI, as it's simply reading BW data. See https://github.com/bwapi/bwapi/blob/master/bwapi/BWAPI/Source/BWAPI/UnitUpdate.cpp#L394 .

We can change the behaviour to be more accurate, especially if everyone has similar DIY solutions. If you guys can post your solutions I could probably come up with something better.

**Comment by albertouri** [comment]
I agree that if you want a specific behavior is better to DIY. But I think that BWAPI isUnderAttack() is pretty useless as it is now (and can lead to misunderstanding to bots programmers). That's why having a "generic" isUnderAttack() can be beneficial for everybody.

I have something similar to krasi0, but maybe we can optimize it on BWAPI. I'm not sure, but I suspect that the unit's order can only be AttackUnit if the target is visible (all other orders like Guard or AttackMove, transition to this one). So probably we can trigger to update the variable when a unit changes to this state.

About the siege tank example. Probably it only change the order to AttackUnit when it finish to siege and start attacking. Then there is no harm to provide this information since when the tank shots it reveals himself from the fog of war.

**Comment by heinermann** [comment]
Can you guys link to your "isUnderAttack" code? I'd like to poop out a new release by Christmas-ish.

**Comment by albertouri** [comment]
Here you can find mine
https://bitbucket.org/auriarte/bwrepdump/src/57a8f2e19fc2acbb9c32cf46ddd1737a5fcb232f/src/CombatTracker.cpp?at=master&fileviewer=file-view-default#CombatTracker.cpp-42

**Comment by heinermann** [comment]
We can try out your neat solution.
